utils/linkcheck.ml

```diff
@@ -0,0 +1,100 @@
+type modname = string
+
+type filename = string
+
+type modname_and_source =
+  { modname  : modname
+  ; filename : filename
+  }
+
+
+module Modname_and_source = struct
+  type t = modname_and_source
+  module Set = Set.Make(struct type nonrec t = t let compare = compare end)
+end
+
+
+
+
+type refs = Modname_and_source.Set.t
+
+type t =
+  { missing_globals : (modname, refs) Hashtbl.t
+  ; provided_globals : (modname, filename) Hashtbl.t
+  ; badly_ordered_deps : (Modname_and_source.t, refs) Hashtbl.t  }
+
+type error =
+  | Missing_implementations of (modname * modname_and_source list) list
+  | Wrong_link_order of (modname_and_source * modname_and_source list) list
+
+let create () =
+  { missing_globals = Hashtbl.create 17
+  ; provided_globals = Hashtbl.create 17
+  ; badly_ordered_deps = Hashtbl.create 17 }
+
+let required t name = Hashtbl.mem t.missing_globals name
+
+let update t k f =
+  let v = try Some (Hashtbl.find t k) with Not_found -> None in
+  Hashtbl.replace t k (f v)
+
+let add_required t by (name : string) =
+  (try
+     let filename = Hashtbl.find t.provided_globals name in
+     update t.badly_ordered_deps {modname = name; filename }
+       (function None -> Modname_and_source.Set.singleton by
+               | Some s -> Modname_and_source.Set.add by s)
+   with Not_found -> ());
+  update t.missing_globals name
+    (function None -> Modname_and_source.Set.singleton by
+            | Some s -> Modname_and_source.Set.add by s)
```

Comment:
```suggestion
let add_required t by (name : string) =
	let add = function 
	  | None -> Modname_and_source.Set.singleton by
	  | Some s -> Modname_and_source.Set.add by s
	in
  (try
     let filename = Hashtbl.find t.provided_globals name in
     update t.badly_ordered_deps {modname = name; filename }
       add
   with Not_found -> ());
  update t.missing_globals name
    add
```

Reply:
If we are being cute with the stdlib,

```ocaml
function 
	  | None -> Modname_and_source.Set.singleton by
	  | Some s -> Modname_and_source.Set.add by s
```

can also be written

```ocaml
fun s -> Modname_and_source.(Set.add by (Option.value s ~default:empty))
```

Reply:
I've took both your suggestion into consideration.